corehq/motech/repeaters/tasks.py

```diff
@@ -322,7 +323,99 @@ def get_repeater_lock(repeater_id):
 
 @task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
 def process_repeater(domain, repeater_id, lock_token):
-    ...
+
+    def get_task_signature(repeat_record):
+        task_ = {
+            State.Pending: process_pending_repeat_record,
+            State.Fail: process_failed_repeat_record,
+        }[repeat_record.state]
+        return task_.s(repeat_record.id, repeat_record.domain)
+
+    repeater = Repeater.objects.get(domain=domain, id=repeater_id)
+    repeat_records = repeater.repeat_records_ready[:repeater.num_workers]
+    header_tasks = [get_task_signature(rr) for rr in repeat_records]
+    chord(header_tasks)(update_repeater.s(repeater_id, lock_token))
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_pending_repeat_record(repeat_record_id, domain):
+    # NOTE: Keep separate from `process_failed_repeat_record()` for
+    # monitoring purposes. `domain` is for tagging in Datadog
+    return process_ready_repeat_record(repeat_record_id)
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_failed_repeat_record(repeat_record_id, domain):
+    # NOTE: Keep separate from `process_pending_repeat_record()` for
+    # monitoring purposes. `domain` is for tagging in Datadog
+    return process_ready_repeat_record(repeat_record_id)
+
+
+def process_ready_repeat_record(repeat_record_id):
+    state_or_none = None
+    with TimingContext('process_repeat_record') as timer:
+        try:
+            repeat_record = (
+                RepeatRecord.objects
+                .prefetch_related('repeater', 'attempt_set')
+                .get(id=repeat_record_id)
+            )
+            report_repeater_attempt(repeat_record.repeater.repeater_id)
+            if not is_repeat_record_ready(repeat_record):
+                return None
+            with timer('fire_timing') as fire_timer:
+                state_or_none = repeat_record.fire(timing_context=fire_timer)
+            report_repeater_usage(
```

Comment:
:thinking: Probably not. @gherceg ?

Reply:
So given request related exceptions are handled in `repeater.fire_for_record` (which gets called by `repeat_record.fire`), it should be true that if none is returned here, it is because an exception was hit when getting the payload or because the repeat record's state is None which I don't think is possible. However it doesn't hurt to report the usage, and we currently _do_ report the timing of both fetching the payload and sending the request.

All that to say, it seems fine to leave it as is for now to remain consistent with how we report timing for the rate limiter. If we decide to reduce the scope of what usage is reported to the request specifically, we can update this code then.